File: BE/src/main/java/com/example/starbucks/repository/CategoryRepository.java
package com.example.starbucks.repository;

import com.example.starbucks.domain.Category;
import com.example.starbucks.domain.CategoryType;
import java.util.Optional;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import java.util.List;

public interface CategoryRepository extends JpaRepository<Category, Long> {

    List<Category> findByType(@Param("type") CategoryType type);

    @Query("select c from Category c "
        + "join fetch c.categoryProducts cp "
        + "join fetch cp.product p "
        + "where c.id = :id "
        + "order by p.orderCount desc ")
    Optional<Category> findByIdWithProduct(@Param("id") Long id);
}

File: Android/app/src/main/java/com/example/starbucks/data/repository/CategoryRepository.kt
package com.example.starbucks.data.repository

import com.example.starbucks.data.vo.CategoryMenuData
import com.example.starbucks.data.vo.DetailCategory

interface CategoryRepository {

    suspend fun getCategory(type: String): List<CategoryMenuData>

    suspend fun getDetailCategory(type: Int): List<DetailCategory>

}